kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DomainMetadataSuite.scala

```diff
@@ -0,0 +1,475 @@
+     * t4 ------- Txn3 starts.
+     * t5 ------- Txn3 commits.
+     * t6 ------------------------ Txn1 commits (SUCCESS or FAIL).
+     */
+    val txn1 = createTxnWithDomainMetadatas(engine, tablePath, currentTxn1DomainMetadatas)
+
+    val txn2 = createTxnWithDomainMetadatas(engine, tablePath, winningTxn2DomainMetadatas)
+    txn2.commit(engine, emptyIterable())
+
+    val txn3 = createTxnWithDomainMetadatas(engine, tablePath, winningTxn3DomainMetadatas)
+    txn3.commit(engine, emptyIterable())
+
+    if (expectedConflict) {
+      // We expect the commit of txn1 to fail because of the conflicting DM actions
+      val ex = intercept[KernelException] {
+        txn1.commit(engine, emptyIterable())
+      }
+      assert(
+        ex.getMessage.contains(
+          "A concurrent writer added a domainMetadata action for the same domain"
+        )
+      )
+    } else {
+      // We expect the commit of txn1 to succeed
+      txn1.commit(engine, emptyIterable())
+      // Verify the final state includes merged domain metadata
+      val expectedMetadata =
+        (winningTxn2DomainMetadatas ++ winningTxn3DomainMetadatas ++ currentTxn1DomainMetadatas)
+          .groupBy(_.getDomain)
+          .mapValues(_.last)
+      assertDomainMetadata(table, engine, expectedMetadata)
+    }
+  }
+
+  test("create table w/o domain metadata") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+
+      // Create an empty table
+      createTxn(engine, tablePath, isNewTable = true, testSchema, Seq.empty)
+        .commit(engine, emptyIterable())
+
+      // Verify that the table doesn't have any domain metadata
+      assertDomainMetadata(table, engine, Map.empty)
+    }
+  }
+
+  test("table w/o domain metadata support fails domain metadata commits") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      // Create an empty table
+      // Its minWriterVersion is 2 and doesn't have 'domainMetadata' in its writerFeatures
+      createTxn(engine, tablePath, isNewTable = true, testSchema, Seq.empty)
+        .commit(engine, emptyIterable())
+
+      val dm1 = new DomainMetadata("domain1", "", false)
+      val txn1 = createTxnWithDomainMetadatas(engine, tablePath, List(dm1))
+
+      // We expect the commit to fail because the table doesn't support domain metadata
+      val e = intercept[KernelException] {
+        txn1.commit(engine, emptyIterable())
+      }
+      assert(
+        e.getMessage
+          .contains(
+            "Cannot commit DomainMetadata action(s) because the feature 'domainMetadata' "
+            + "is not supported on this table."
+          )
+      )
+
+      // Set writer version and writer feature to support domain metadata
+      setDomainMetadataSupport(engine, tablePath)
+
+      // Commit domain metadata again and expect success
+      val txn2 = createTxnWithDomainMetadatas(engine, tablePath, List(dm1))
+      txn2.commit(engine, emptyIterable())
```

Comment:
nit but can we use the helper function that also checks the committed DM here?

Reply:
Yes we should use that. Updated.